Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 2022 Apple Inc.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "config.h"
+#include "GridMasonryLayout.h"
+
+#include "rendering/style/GridPositionsResolver.h"
+
+namespace WebCore {
+
+void GridMasonryLayout::performMasonryPlacement(const HashMap<RenderBox*, GridArea>& firstTrackItems, const Vector<RenderBox*> itemsWithDefiniteGridAxisPosition, const Vector<RenderBox*> itemsWithIndefinitePosition, GridTrackSizingDirection masonryAxisDirection)
+{
+    m_masonryAxisDirection = masonryAxisDirection;
+    m_gridAxisDirection = m_masonryAxisDirection == ForRows ? ForColumns : ForRows;
+    m_masonryAxisGridGap = m_renderGrid.gridGap(m_masonryAxisDirection);
+    m_gridAxisTracksCount = m_renderGrid.numTracks(m_gridAxisDirection);
+    m_runningPositions = Vector<LayoutUnit>(m_gridAxisTracksCount);
```

Comment:
If we tend to have lots of tracks, we could consider resizing the existing object so it could resize-in place if the grid was undergoing a second round of layout.

Reply:
Done!